### PR TITLE
Remove HybridDictionary

### PR DESCRIPTION
**Pull request by ccastanedaucf**
While I don't have microbenchmarks, I do have profiling that shows every usage of ```HybridDictionary``` is significantly slower than using ```Collections.Generic.Dictionary``` in the same location, including when the dictionary size is typically small enough for ```HybridDictionary``` to remain a ```List``` such as ```TaskFactoryWrapper```. Mostly slows down item metadata operations during evaluation.

image

_(edited second trace since the original capture got skewed by the other methods in ```TaskFactoryWrapper```_
image

### CODE REVIEW

**Comment by AndyGerlicher** [comment]
Wow, looks good. I don't trust the memory stuff really at all, wish I did :(.

But 12% on the dotnet console project is crazy!

### DesignTimeBuild Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 34.2357 -> 29.8414 (-12.835%)
DotnetWebProject | :white_check_mark: | yes | 228.5297 -> 208.2002 (-8.896%)
DotnetMvcProject | :white_check_mark: | yes | 230.8673 -> 206.4007 (-10.598%)
Picasso | :white_check_mark: | yes | 1514.7158 -> 1472.1144 (-2.813%)
SmallP2POldCsproj | :white_check_mark: | yes | 66.5477 -> 64.6826 (-2.803%)
SmallP2PNewCsproj | :white_check_mark: | yes | 604.6303 -> 576.4128 (-4.667%)
LargeP2POldCsproj | :white_check_mark: | yes | 10780.354 -> 10606.0966 (-1.616%)
OrchardCore | :white_check_mark: | yes | 45335.7867 -> 43666.7994 (-3.681%)

### DesignTimeBuild Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 6243427 -> 6406595 (2.613%)
DotnetWebProject | :ok_hand: | no | 21068659 -> 22245748 (5.587%)
DotnetMvcProject | :red_circle: | yes | 20698244 -> 24322506 (17.51%)
Picasso | :ok_hand: | no | 173596498 -> 176823351 (1.859%)
SmallP2POldCsproj | :red_circle: | yes | 8421434 -> 8514299 (1.103%)
SmallP2PNewCsproj | :ok_hand: | no | 91596634 -> 92115117 (0.566%)
LargeP2POldCsproj | :white_check_mark: | yes | 1104324896 -> 1023254636 (-7.341%)
OrchardCore | :white_check_mark: | yes | 2713724824 -> 2529473041 (-6.79%)

### SerialEvaluationIsolatedContext Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 33.2709 -> 32.996 (-0.826%)
DotnetWebProject | :white_check_mark: | yes | 43.3831 -> 43.2131 (-0.392%)
DotnetMvcProject | ::ok_hand: | no | 48.6875 -> 48.6157 (-0.147%)
Picasso | :red_circle: | yes | 293.0064 -> 297.8211 (1.643%)
SmallP2POldCsproj | :white_check_mark: | yes | 50.2931 -> 49.9456 (-0.691%)
SmallP2PNewCsproj | ::ok_hand: | no | 188.7995 -> 188.2542 (-0.289%)
LargeP2POldCsproj | :ok_hand: | no | 852.7351 -> 852.8619 (0.015%)
OrchardCore | ::ok_hand: | no | 2902.7991 -> 2900.0241 (-0.096%)

### SerialEvaluationIsolatedContext Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 5612480 -> 5644864 (0.577%)
DotnetWebProject | :red_circle: | yes | 7477654 -> 7519920 (0.565%)
DotnetMvcProject | :red_circle: | yes | 8087893 -> 8122381 (0.426%)
Picasso | :red_circle: | yes | 10025603 -> 10295282 (2.69%)
SmallP2POldCsproj | :red_circle: | yes | 6856971 -> 6902614 (0.666%)
SmallP2PNewCsproj | :ok_hand: | no | 13107652 -> 14150019 (7.952%)
LargeP2POldCsproj | :ok_hand: | no | 34207714 -> 34772916 (1.652%)
OrchardCore | ::ok_hand: | no | 70961891 -> 70244829 (-1.01%)

### SerialEvaluationSharedContextSecondRun Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :ok_hand: | no | 29.4819 -> 29.5116 (0.101%)
DotnetWebProject | :white_check_mark: | yes | 36.897 -> 36.6572 (-0.65%)
DotnetMvcProject | :white_check_mark: | yes | 38.967 -> 38.7685 (-0.509%)
Picasso | :white_check_mark: | yes | 214.3561 -> 213.913 (-0.207%)
SmallP2POldCsproj | :ok_hand: | no | 40.6895 -> 40.7016 (0.03%)
SmallP2PNewCsproj | :white_check_mark: | yes | 129.4603 -> 126.7616 (-2.085%)
LargeP2POldCsproj | :red_circle: | yes | 633.8486 -> 644.1612 (1.627%)
Generated_100_100_v150 | :white_check_mark: | yes | 1148.72 -> 1140.6148 (-0.706%)
OrchardCore | :red_circle: | yes | 1511.903 -> 1522.2263 (0.683%)
Roslyn | :red_circle: | yes | 2528.6508 -> 2560.8814 (1.275%)
WebLargeCore | :red_circle: | yes | 1762.9418 -> 1775.9184 (0.736%)

### SerialEvaluationSharedContextSecondRun Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 5518568 -> 5551265 (0.592%)
DotnetWebProject | :red_circle: | yes | 6842353 -> 6883020 (0.594%)
DotnetMvcProject | :red_circle: | yes | 7276177 -> 7309270 (0.455%)
Picasso | :red_circle: | yes | 34809919 -> 34843759 (0.097%)
SmallP2POldCsproj | :red_circle: | yes | 6837007 -> 6870015 (0.483%)
SmallP2PNewCsproj | :red_circle: | yes | 22382378 -> 22433121 (0.227%)
LargeP2POldCsproj | :white_check_mark: | yes | 31267502 -> 30089451 (-3.768%)
Generated_100_100_v150 | :white_check_mark: | yes | 42903962 -> 42869261 (-0.081%)
OrchardCore | :ok_hand: | no | 76620068 -> 77708470 (1.421%)
Roslyn | ::ok_hand: | no | 79289245 -> 79223100 (-0.083%)
WebLargeCore | ::ok_hand: | no | 65143722 -> 64533208 (-0.937%)

### SerialEvaluationSharedContext Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 32.4741 -> 32.456 (-0.056%)
DotnetWebProject | :white_check_mark: | yes | 42.5751 -> 42.3789 (-0.461%)
DotnetMvcProject | :white_check_mark: | yes | 47.9576 -> 47.8234 (-0.28%)
Picasso | :white_check_mark: | yes | 243.0388 -> 241.192 (-0.76%)
SmallP2POldCsproj | :white_check_mark: | yes | 44.1976 -> 44.0032 (-0.44%)
SmallP2PNewCsproj | :white_check_mark: | yes | 157.6546 -> 155.8848 (-1.123%)
LargeP2POldCsproj | :red_circle: | yes | 676.9174 -> 688.0718 (1.648%)
Generated_100_100_v150 | :white_check_mark: | yes | 1169.0193 -> 1154.3772 (-1.253%)
OrchardCore | :ok_hand: | no | 1870.9 -> 1878.8256 (0.424%)
Roslyn | :red_circle: | yes | 2892.53 -> 2919.2293 (0.923%)
WebLargeCore | :red_circle: | yes | 2043.3093 -> 2059.1457 (0.775%)

### SerialEvaluationSharedContext Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 5682607 -> 5715112 (0.572%)
DotnetWebProject | :red_circle: | yes | 7530252 -> 7576983 (0.621%)
DotnetMvcProject | :red_circle: | yes | 8134073 -> 8169379 (0.434%)
Picasso | :ok_hand: | no | 10229767 -> 10284655 (0.537%)
SmallP2POldCsproj | :red_circle: | yes | 6870448 -> 6911307 (0.595%)
SmallP2PNewCsproj | :ok_hand: | no | 11919168 -> 12434157 (4.321%)
LargeP2POldCsproj | :ok_hand: | no | 32004187 -> 32018178 (0.044%)
Generated_100_100_v150 | :red_circle: | yes | 43350055 -> 43567548 (0.502%)
OrchardCore | :ok_hand: | no | 60930170 -> 64525070 (5.9%)
Roslyn | :ok_hand: | no | 82687293 -> 83452758 (0.926%)
WebLargeCore | ::ok_hand: | no | 69554480 -> 67521038 (-2.924%)

File: src/Tasks/FindReferenceAssembliesForReferences.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (ccastanedaucf) [inline src/Tasks/FindReferenceAssembliesForReferences.cs]: Ignore this file, totally forgot to remove it when I was testing something with item metadata

**Comment by AndyGerlicher** [comment]
Seems better I think?

### DesignTimeBuild Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 33.8947 -> 29.7468 (-12.238%)
DotnetWebProject | :white_check_mark: | yes | 228.2893 -> 207.6751 (-9.03%)
DotnetMvcProject | :white_check_mark: | yes | 228.6045 -> 208.6962 (-8.709%)
Picasso | :white_check_mark: | yes | 1507.178 -> 1458.8566 (-3.206%)
SmallP2POldCsproj | :white_check_mark: | yes | 65.8393 -> 64.5243 (-1.997%)
SmallP2PNewCsproj | :white_check_mark: | yes | 608.7546 -> 572.965 (-5.879%)
LargeP2POldCsproj | :white_check_mark: | yes | 10910.6633 -> 10703.4772 (-1.899%)
OrchardCore | :white_check_mark: | yes | 45275.3478 -> 43652.4022 (-3.585%)

### DesignTimeBuild Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 6242393 -> 6406064 (2.622%)
DotnetWebProject | :ok_hand: | no | 20344546 -> 22151428 (8.881%)
DotnetMvcProject | :ok_hand: | no | 21652961 -> 21803781 (0.697%)
Picasso | :white_check_mark: | yes | 186021500 -> 176484027 (-5.127%)
SmallP2POldCsproj | :red_circle: | yes | 8420661 -> 8527431 (1.268%)
SmallP2PNewCsproj | ::ok_hand: | no | 89570468 -> 87886192 (-1.88%)
LargeP2POldCsproj | :white_check_mark: | yes | 1081144997 -> 1004956666 (-7.047%)
OrchardCore | :white_check_mark: | yes | 2581202228 -> 2480195376 (-3.913%)

### SerialEvaluationIsolatedContext Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 33.3349 -> 33.0074 (-0.982%)
DotnetWebProject | :white_check_mark: | yes | 43.4766 -> 43.0258 (-1.037%)
DotnetMvcProject | :ok_hand: | no | 48.7474 -> 48.8115 (0.131%)
Picasso | :ok_hand: | no | 294.1861 -> 295.7546 (0.533%)
SmallP2POldCsproj | :white_check_mark: | yes | 49.7431 -> 49.4038 (-0.682%)
SmallP2PNewCsproj | :white_check_mark: | yes | 190.1567 -> 187.4466 (-1.425%)
LargeP2POldCsproj | :red_circle: | yes | 867.8351 -> 870.549 (0.313%)
OrchardCore | ::ok_hand: | no | 2916.9686 -> 2907.642 (-0.32%)

### SerialEvaluationIsolatedContext Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 5612813 -> 5645658 (0.585%)
DotnetWebProject | :red_circle: | yes | 7480734 -> 7520175 (0.527%)
DotnetMvcProject | :red_circle: | yes | 8090148 -> 8125948 (0.443%)
Picasso | ::ok_hand: | no | 12701476 -> 11920880 (-6.146%)
SmallP2POldCsproj | :red_circle: | yes | 6857426 -> 6903680 (0.675%)
SmallP2PNewCsproj | ::ok_hand: | no | 14251663 -> 13480528 (-5.411%)
LargeP2POldCsproj | :ok_hand: | no | 33014267 -> 33709064 (2.105%)
OrchardCore | ::ok_hand: | no | 72304549 -> 70959570 (-1.86%)

### SerialEvaluationSharedContextSecondRun Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 29.6204 -> 29.4554 (-0.557%)
DotnetWebProject | :white_check_mark: | yes | 37.0008 -> 36.8028 (-0.535%)
DotnetMvcProject | :white_check_mark: | yes | 39.1301 -> 38.8278 (-0.773%)
Picasso | :white_check_mark: | yes | 216.5242 -> 213.7179 (-1.296%)
SmallP2POldCsproj | :white_check_mark: | yes | 41.2118 -> 40.8506 (-0.876%)
SmallP2PNewCsproj | :white_check_mark: | yes | 129.9116 -> 127.1918 (-2.094%)
LargeP2POldCsproj | :red_circle: | yes | 639.3154 -> 648.4657 (1.431%)
Generated_100_100_v150 | :red_circle: | yes | 1127.8622 -> 1168.3364 (3.589%)
OrchardCore | ::ok_hand: | no | 1510.8593 -> 1510.7587 (-0.007%)
Roslyn | :white_check_mark: | yes | 2541.3993 -> 2518.5333 (-0.9%)
WebLargeCore | :white_check_mark: | yes | 1790.1249 -> 1779.368 (-0.601%)

### SerialEvaluationSharedContextSecondRun Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 5518541 -> 5551415 (0.596%)
DotnetWebProject | :red_circle: | yes | 6842450 -> 6883034 (0.593%)
DotnetMvcProject | :red_circle: | yes | 7275926 -> 7309335 (0.459%)
Picasso | :ok_hand: | no | 34086176 -> 34844055 (2.223%)
SmallP2POldCsproj | :red_circle: | yes | 6837109 -> 6870087 (0.482%)
SmallP2PNewCsproj | :red_circle: | yes | 22382522 -> 22433744 (0.229%)
LargeP2POldCsproj | :ok_hand: | no | 30649688 -> 30711374 (0.201%)
Generated_100_100_v150 | :white_check_mark: | yes | 42927539 -> 42873821 (-0.125%)
OrchardCore | ::ok_hand: | no | 74608906 -> 74359412 (-0.334%)
Roslyn | :ok_hand: | no | 79455600 -> 80854478 (1.761%)
WebLargeCore | :red_circle: | yes | 65074536 -> 65227529 (0.235%)

### SerialEvaluationSharedContext Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 32.5538 -> 32.3754 (-0.548%)
DotnetWebProject | ::ok_hand: | no | 42.5435 -> 42.4806 (-0.148%)
DotnetMvcProject | :white_check_mark: | yes | 48.107 -> 47.9752 (-0.274%)
Picasso | :white_check_mark: | yes | 244.3304 -> 241.788 (-1.041%)
SmallP2POldCsproj | :white_check_mark: | yes | 44.5399 -> 44.3172 (-0.5%)
SmallP2PNewCsproj | :white_check_mark: | yes | 158.5248 -> 155.4807 (-1.92%)
LargeP2POldCsproj | :red_circle: | yes | 681.307 -> 693.0567 (1.725%)
Generated_100_100_v150 | :red_circle: | yes | 1146.6865 -> 1182.882 (3.157%)
OrchardCore | :ok_hand: | no | 1862.5778 -> 1864.3182 (0.093%)
Roslyn | :white_check_mark: | yes | 2912.6683 -> 2883.5968 (-0.998%)
WebLargeCore | ::ok_hand: | no | 2067.9426 -> 2063.7004 (-0.205%)

### SerialEvaluationSharedContext Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 5682393 -> 5715289 (0.579%)
DotnetWebProject | :red_circle: | yes | 7527978 -> 7578022 (0.665%)
DotnetMvcProject | :red_circle: | yes | 8132947 -> 8170823 (0.466%)
Picasso | ::ok_hand: | no | 10229650 -> 9563764 (-6.509%)
SmallP2POldCsproj | :red_circle: | yes | 6870410 -> 6911292 (0.595%)
SmallP2PNewCsproj | :red_circle: | yes | 10451273 -> 12610915 (20.664%)
LargeP2POldCsproj | :ok_hand: | no | 31372564 -> 32011553 (2.037%)
Generated_100_100_v150 | :red_circle: | yes | 43436044 -> 43557990 (0.281%)
OrchardCore | :ok_hand: | no | 63096085 -> 65891051 (4.43%)
Roslyn | :ok_hand: | no | 81835188 -> 82701432 (1.059%)
WebLargeCore | :ok_hand: | no | 66212012 -> 67873517 (2.509%)

File: src/Build/Instance/TaskFactoryWrapper.cs [hunk 2]
```diff
@@ -275,27 +274,43 @@ private void PopulatePropertyInfoCacheIfNecessary()
...
+                        {
+                            _namesOfPropertiesWithOutputAttribute = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+                        }
+
                         // we have a output attribute defined, keep a record of that
                         _namesOfPropertiesWithOutputAttribute[propertyInfo.Name] = String.Empty;
                     }
                 }
 
                 // Toss the dictionaries if we can as often they are empty (at least the last three are)
```
Comment (rainersigwald) [inline src/Build/Instance/TaskFactoryWrapper.cs]: Remove this comment too?

**Comment by rainersigwald** [comment]
Not sure what the deal is with the wip-bot; I'll remove it as a required check.

**Comment by AndyGerlicher** [comment]
Just FYI, I quickly implemented the single -> Dictionary hybrid... it seems worse:

_\msbuild.A is HybridDictionaryNoList, \msbuild.B is this PR_
### DesignTimeBuild Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 30.294 -> 29.6722 (-2.053%)
DotnetWebProject | ::ok_hand: | no | 208.2015 -> 207.8042 (-0.191%)
DotnetMvcProject | ::ok_hand: | no | 210.7119 -> 210.4125 (-0.142%)
Picasso | :white_check_mark: | yes | 1466.5504 -> 1462.3075 (-0.289%)
SmallP2POldCsproj | :white_check_mark: | yes | 64.9987 -> 64.7198 (-0.429%)
SmallP2PNewCsproj | ::ok_hand: | no | 574.9533 -> 574.1841 (-0.134%)
LargeP2POldCsproj | :red_circle: | yes | 10622.8903 -> 10664.9453 (0.396%)
OrchardCore | :white_check_mark: | yes | 43889.4783 -> 43634.7676 (-0.58%)

### DesignTimeBuild Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 6398765 -> 6410133 (0.178%)
DotnetWebProject | :white_check_mark: | yes | 24750483 -> 20811793 (-15.914%)
DotnetMvcProject | ::ok_hand: | no | 21826926 -> 21094491 (-3.356%)
Picasso | :ok_hand: | no | 177495976 -> 180126811 (1.482%)
SmallP2POldCsproj | :red_circle: | yes | 8462597 -> 8506089 (0.514%)
SmallP2PNewCsproj | :ok_hand: | no | 91899470 -> 92019530 (0.131%)
LargeP2POldCsproj | :white_check_mark: | yes | 1068280894 -> 1024618641 (-4.087%)
OrchardCore | :white_check_mark: | yes | 2595863447 -> 2516052658 (-3.075%)

I think it's the inefficient iterators and such. When there are single items it's creating arrays and likely boxing the KVP. Maybe we could optimize it further, but I vote we just merge this since it's very nice gains!

**Pull request merged**